src/Microsoft.OData.Core/ODataInnerError.cs

```diff
@@ -33,38 +41,71 @@ public ODataInnerError(Exception exception)
         {
             ExceptionUtils.CheckArgumentNotNull(exception, "exception");
 
-            this.Message = exception.Message ?? string.Empty;
-            this.TypeName = exception.GetType().FullName;
-            this.StackTrace = exception.StackTrace;
-
             if (exception.InnerException != null)
             {
                 this.InnerError = new ODataInnerError(exception.InnerException);
             }
+
+            Properties = new Dictionary<string, ODataValue>();
+
+            //// NOTE: we add empty elements if no information is provided for the message, error type and stack trace
+            ////       to stay compatible with Astoria.
+            Properties.Add(JsonConstants.ODataErrorInnerErrorMessageName, exception.Message.ToODataValue() ?? new ODataNullValue());
+            Properties.Add(JsonConstants.ODataErrorInnerErrorTypeNameName, exception.GetType().FullName.ToODataValue() ?? new ODataNullValue());
+            Properties.Add(JsonConstants.ODataErrorInnerErrorStackTraceName, exception.StackTrace.ToODataValue() ?? new ODataNullValue());
+        }
+
+        /// <summary>
+        ///  Initializes a new instance of the <see cref="T:Microsoft.OData.ODataInnerError" /> class with a dictionary of property names and corresponding ODataValues.
+        /// </summary>
+        /// <param name="properties">Dictionary of string keys with ODataValue as value. Key string indicates the property name where as the value of the property is encapsulated in ODataValue.</param>
+        public ODataInnerError(IDictionary<string, ODataValue> properties)
+        {
+            ExceptionUtils.CheckArgumentNotNull(properties, "properties");
+
+            Properties = properties;
+        }
+
+        /// <summary>
+        /// Creates an instance of ODataInnerError with an empty property bag.
+        /// </summary>
+        /// <returns>returns an instance of ODataInnerError with empty property bag.</returns>
+        public static ODataInnerError CreateEmptyInnerError()

```

Comment:
>CreateEmptyInnerError [](start = 38, length = 21)

Also note that, by spec, the ODataError must have non-empty "Message" and "Code" strings.

Reply:
I am modifying ODataInnerError here which does not have such constraints. For ODataError, we currently do not enforce such constraints in the implementation and we cannot even do it without having a breaking change. 

---
In reply to: [321910714](https://github.com/OData/odata.net/pull/1523#discussion_r321910714) [](ancestors = 321910714)